caliban_toolbox/reshape_data.py

```diff
@@ -387,3 +387,283 @@ def save_npzs_for_caliban(resized_xr, original_xr, log_data,  save_dir, blank_la
+    fov_names = log_data['fov_names']
+    # combine all npz crops into a single stack
+    crop_stack = load_npzs(crop_dir=crop_dir, log_data=log_data, verbose=verbose)
+
+    # stitch crops together into single contiguous image
+    stitched_images = stitch_crops(annotated_data=crop_stack, log_data=log_data)
+
+    # crop image down to original size
+    if row_padding > 0:
+        stitched_images = stitched_images[:, :, :, :, :-row_padding, :, :]
+    if col_padding > 0:
+        stitched_images = stitched_images[:, :, :, :, :, :-col_padding, :]
+
+    _, stack_len, _, _, row_len, col_len, _ = log_data['original_shape']
+    stitched_xr = xr.DataArray(data=stitched_images,
+                               coords=[fov_names, range(stack_len), range(1), range(1), range(row_len), range(col_len),
+                                       ['segmentation_label']], dims=['fovs', 'stacks', 'crops', 'slices',
+                                                                      'rows', 'cols', 'channels'])
+
+    stitched_xr.to_netcdf(os.path.join(crop_dir, 'stitched_images.nc'))
+
+
+def stitch_slices(slice_stack, log_data):
+    """Helper function to stitch slices together back into original sized array
+
+    Args:
+        slice_stack: xarray of shape [fovs, stacks, crops, slices, rows, cols, segmentation_label]
+        log_data: log data produced from creation of slice stack
+
+    Returns:
+        stitched_slices: xarray of shape [fovs, stacks, crops, slices, rows, cols, segmentation_label]
+    """
+
+    # get parameters from dict
+    fov_len, stack_len, crop_num, _, row_len, col_len, chan_len = log_data['original_shape']
+    crop_num = log_data.get('num_crops', crop_num)
+    row_len = log_data.get('row_crop_size', row_len)
+    col_len = log_data.get('col_crop_size', col_len)
+
+    slice_start_indices, slice_end_indices = log_data['slice_start_indices'], log_data['slice_end_indices']
+    num_slices, fov_names = log_data['num_slices'], log_data['fov_names']
+
+    stitched_slices = np.zeros((fov_len, stack_len, crop_num, 1, row_len, col_len, 1))
+
+    # loop slice indices to generate sliced data
+    for i in range(num_slices - 1):
+        stitched_slices[:, slice_start_indices[i]:slice_end_indices[i], :, 0, ...] = slice_stack[:, :, :, i, ...]
+
+    # last slice, only index into stack the amount two indices are separated
+    last_idx = num_slices - 1
+    slice_len = slice_end_indices[last_idx] - slice_start_indices[last_idx]
+    stitched_slices[:, slice_start_indices[last_idx]:slice_end_indices[last_idx], :, 0, ...] = slice_stack[:, :slice_len, :, last_idx, ...]
+
+    stitched_xr = xr.DataArray(stitched_slices,
+                               coords=[fov_names, range(stack_len), range(crop_num), range(1), range(row_len),
+                                       range(col_len), ['segmentation_label']],
+                               dims=['fovs', 'stacks', 'crops', 'slices', 'rows', 'cols', 'channels'])
```

Comment:
Here too I think a couple extra lines could make this a bit easier to understand.

Reply:
Agreed, new changes make this much cleaner! Good catch